toolchain/parse/typed_nodes_test.cpp

```diff
@@ -0,0 +1,133 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/parse/typed_nodes.h"
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include <forward_list>
+
+#include "toolchain/diagnostics/mocks.h"
+#include "toolchain/lex/lex.h"
+#include "toolchain/lex/tokenized_buffer.h"
+#include "toolchain/parse/extract_file.h"
+#include "toolchain/parse/tree.h"
+
+namespace Carbon::Parse {
+namespace {
+
+// Check that each node kind defines a Kind member using the correct
+// NodeKind enumerator.
+#define CARBON_PARSE_NODE_KIND(Name) \
+  static_assert(Name::Kind == NodeKind::Name, #Name);
+#include "toolchain/parse/node_kind.def"
+
+class TypedNodeTest : public ::testing::Test {
+ protected:
+  auto GetSourceBuffer(llvm::StringRef t) -> SourceBuffer& {
+    CARBON_CHECK(fs_.addFile("test.carbon", /*ModificationTime=*/0,
+                             llvm::MemoryBuffer::getMemBuffer(t)));
+    source_storage_.push_front(std::move(
+        *SourceBuffer::CreateFromFile(fs_, "test.carbon", consumer_)));
+    return source_storage_.front();
+  }
+
+  auto GetTokenizedBuffer(llvm::StringRef t) -> Lex::TokenizedBuffer& {
+    token_storage_.push_front(
+        Lex::Lex(value_stores_, GetSourceBuffer(t), consumer_));
+    return token_storage_.front();
+  }
+
+  auto GetTree(llvm::StringRef t) -> Tree& {
+    tree_storage_.push_front(Tree::Parse(GetTokenizedBuffer(t), consumer_,
+                                         /*vlog_stream=*/nullptr));
+    return tree_storage_.front();
+  }
+
+  SharedValueStores value_stores_;
+  llvm::vfs::InMemoryFileSystem fs_;
+  std::forward_list<SourceBuffer> source_storage_;
+  std::forward_list<Lex::TokenizedBuffer> token_storage_;
+  std::forward_list<Tree> tree_storage_;
+  DiagnosticConsumer& consumer_ = ConsoleDiagnosticConsumer();
+};
+
+TEST_F(TypedNodeTest, Empty) {
+  auto* tree = &GetTree("");
+  auto file = File::Make(tree);
```

Comment:
I think we should rename `File::Make` to `File::Extract` for symmetry with other extraction operations. I also wonder if we can switch this around to being `tree->ExtractFile()`, but perhaps the dependencies there are awkward?